File: common/python/t5_testkit/layers/attention.py
import torch
import torch.nn as nn
import numpy as np


def relative_position_bucket(relative_position, bidirectional, num_buckets, max_distance=128):
    rp = relative_position

    ret = 0
    if bidirectional:
        num_buckets //= 2
        ret += (rp > 0).long() * num_buckets
        rp = rp.abs()
    else:
        rp = torch.clamp(-rp, min=0)

    max_exact = num_buckets // 2
    is_small = rp < max_exact

    large_pos = max_exact + (
        torch.log(rp.float() / max_exact) /
        np.log(max_distance / max_exact) *
        (num_buckets - max_exact)
    ).long()

    large_pos = torch.clamp(large_pos, max=num_buckets - 1)

    ret += torch.where(is_small, rp, large_pos)
    return ret


class T5Attention(nn.Module):

    def __init__(self, config, has_relative_bias=True):
        super().__init__()

        self.d_model = config["d_model"]
        self.d_kv = config["d_kv"]
        self.n_heads = config["num_heads"]
        self.inner_dim = self.n_heads * self.d_kv

        self.num_buckets = config["relative_attention_num_buckets"]
        self.bidirectional = not config.get("is_decoder", False)

        self.has_relative_bias = has_relative_bias

        self.q = nn.Linear(self.d_model, self.inner_dim, bias=False)
        self.k = nn.Linear(self.d_model, self.inner_dim, bias=False)
        self.v = nn.Linear(self.d_model, self.inner_dim, bias=False)
        self.o = nn.Linear(self.inner_dim, self.d_model, bias=False)

        if has_relative_bias:
            self.relative_attention_bias = nn.Embedding(self.num_buckets, self.n_heads)

    def compute_bias(self, q_len, k_len, device):
        q_pos = torch.arange(q_len, device=device)[:, None]
        k_pos = torch.arange(k_len, device=device)[None, :]
        rel = k_pos - q_pos

        buckets = relative_position_bucket(
            rel,
            self.bidirectional,
            self.num_buckets
        ) 

        bias = self.relative_attention_bias(buckets)

        return bias.permute(2, 0, 1).unsqueeze(0)

    def forward(self, hidden_states, key_value_states=None):
        B, S, _ = hidden_states.shape

        kv = hidden_states if key_value_states is None else key_value_states
        K_len = kv.shape[1]

        q = self.q(hidden_states) 
        k = self.k(kv)
        v = self.v(kv)

        q = q.view(B, S, self.n_heads, self.d_kv).transpose(1, 2)
        k = k.view(B, K_len, self.n_heads, self.d_kv).transpose(1, 2)
        v = v.view(B, K_len, self.n_heads, self.d_kv).transpose(1, 2)

        scores = torch.matmul(q, k.transpose(-2, -1))

        if self.has_relative_bias:
            pos_bias = self.compute_bias(S, K_len, hidden_states.device)
        else:
            pos_bias = torch.zeros((1, self.n_heads, S, K_len),
                                   device=hidden_states.device)

        scores = scores + pos_bias

        attn = torch.softmax(scores, dim=-1)
        out = torch.matmul(attn, v)

        out = out.transpose(1, 2).reshape(B, S, self.inner_dim)
        out = self.o(out)

        return out, pos_bias

File: version1_plainC/src/attention.cpp
#include "attention.hpp"
#include <algorithm>
#include <cmath>
#include <cstring>
#include "config.hpp"

MultiHeadAttention::MultiHeadAttention(bool has_bias, bool is_decoder)
    : q_proj(T5Config::d_model, T5Config::num_heads * T5Config::d_kv),
      k_proj(T5Config::d_model, T5Config::num_heads * T5Config::d_kv),
      v_proj(T5Config::d_model, T5Config::num_heads * T5Config::d_kv),
      o_proj(T5Config::num_heads * T5Config::d_kv, T5Config::d_model),
      d_model(T5Config::d_model),
      n_heads(T5Config::num_heads),
      d_kv(T5Config::d_kv),
      inner_dim(T5Config::num_heads * T5Config::d_kv),
      has_relative_bias(has_bias),
      is_decoder(is_decoder)
{
    if (has_relative_bias)
        relative_attention_bias = Tensor({T5Config::num_buckets, n_heads});
}

// Reference: "Exploring the Limits of Transfer Learning with a Unified Text-to-Text Transformer"
// https://arxiv.org/abs/1910.10683
// https://colab.research.google.com/drive/19WgoxCSzAnzi0MueBRP6KY99Ewre1FFy?usp=sharing#scrollTo=W4zmUH7Etndk

int MultiHeadAttention::relative_position_to_bucket(int relative_position)
{
    int number_buckets = T5Config::num_buckets;
    int bucket = 0;

    // In the bidirectional Encoder, we can see both directions of our input
    //(previous words and next words), so the buckets are split.
    if (!is_decoder)
    {
        number_buckets = number_buckets / 2;

        // If key > query this is forward direction
        if (relative_position > 0)
            bucket += number_buckets;

        relative_position = std::abs(relative_position);
    }
    else
    {
        // The Decoder only uses past positions, so the relative position
        // (key - query) is always negative or zero.
        // We negate it to get a positive distance.
        relative_position = -std::min(relative_position, 0);
    }

    int max_exact = number_buckets / 2;

    // if is small -> exact buckets
    if (relative_position < max_exact)
        return bucket + relative_position;

    // if is large: log buckets
    float log_ratio = std::log((float)relative_position / max_exact);
    float log_max = std::log((float)T5Config::max_distance / max_exact);

    float pos = max_exact + (log_ratio / log_max) * (num_buckets - max_exact);
    bucket += std::min((int)pos, number_buckets - 1);

    return bucket;
}

void MultiHeadAttention::compute_relative_position_bias(
    Tensor &bias, int query_len, int key_len)
{
    if (!has_relative_bias)
    {
        bias = Tensor({bias.size()});
        return;
    }

    for (int head = 0; head < n_heads; ++head)
    {
        for (int query = 0; query < query_len; ++query)
        {
            for (int key = 0; key < key_len; ++key)
            {

                int relative_position = key - query;
                int bucket_idx = relative_position_to_bucket(relative_position);

                int index = head * (query_len * key_len) + query * key_len + key;

                bias.data[index] =
                    relative_attention_bias.data[bucket_idx * n_heads + head];
            }
        }
    }
}
std::pair<Tensor, Tensor> MultiHeadAttention::forward(
    const Tensor &hidden_states,
    const Tensor *key_value_states,
    const Tensor *position_bias)
{

    int seq_len = hidden_states.shape[0];

    Tensor q = q_proj.forward(hidden_states);

    // key_value_states are used for cross-attention (decoder)
    // hidden_states are used for self-attention (encoder/decoder)
    const Tensor &kv_in = key_value_states ? *key_value_states : hidden_states;
    Tensor k = k_proj.forward(kv_in);
    Tensor v = v_proj.forward(kv_in);

    int k_len = kv_in.shape[0];

    // [Seq, Heads*Dim] -> [Heads, Seq, Dim]
    q = q.reshape({seq_len, n_heads, d_kv}).permute({1, 0, 2});
    k = k.reshape({k_len, n_heads, d_kv}).permute({1, 0, 2});
    v = v.reshape({k_len, n_heads, d_kv}).permute({1, 0, 2});

    Tensor bias({n_heads, seq_len, k_len});
    if (position_bias)
    {
        bias = *position_bias;
    }
    else
    {
        compute_relative_position_bias(bias, seq_len, k_len);
    }

    Tensor context_layer({n_heads, seq_len, d_kv});

    int q_head_size = seq_len * d_kv;
    int k_head_size = k_len * d_kv;
    int bias_head_size = seq_len * k_len;

    for (int h = 0; h < n_heads; ++h)
    {

        Tensor q_h({seq_len, d_kv});
        memcpy(q_h.data.data(),
               q.data.data() + h * q_head_size,
               q_head_size * sizeof(float));

        Tensor k_h({k_len, d_kv});
        memcpy(k_h.data.data(),
               k.data.data() + h * k_head_size,
               k_head_size * sizeof(float));

        Tensor v_h({k_len, d_kv});
        memcpy(v_h.data.data(),
               v.data.data() + h * k_head_size,
               k_head_size * sizeof(float));

        // Attention(Q,K,V) = Softmax((Q * K^T ) + B)*V

        Tensor scores = q_h.matmul(k_h.transpose());

        int bias_offset = h * bias_head_size;

        for (int i = 0; i < bias_head_size; ++i)
        {
            scores.data[i] += bias.data[bias_offset + i];
        }

        scores = scores.softmax();

        Tensor head_out = scores.matmul(v_h);

        memcpy(context_layer.data.data() + h * q_head_size,
               head_out.data.data(),
               q_head_size * sizeof(float));
    }

    // [n_heads, seq, d_kv] -> [seq, n_heads, d_kv] -> [seq, inner_dim]
    context_layer = context_layer.permute({1, 0, 2})
                        .reshape({seq_len, inner_dim});

    Tensor output = o_proj.forward(context_layer);

    return {output, bias};
}